### DISCUSSION TITLE
Not able add terms with children property in a framework

### DISCUSSION THREAD

**Discussion by ramkumar-ns-dev** [Sunbird-Knowlg/Community#180] [Bugs]
### **Prerequisites**
I have searched for existing issues that already report this problem, without success.

### **Current Behaviour**
I'm trying to add terms in framework with children properties, it should return success status. But it's throwing 500 internal server error. 

### **Expected Behaviour**
While the user adding the terms in framework using "terms create" API end point, It should return success status.

### **Steps to Reproduce**
	1. get the framework Identifier.
	2. Create category using framework ID. using following End point https://dev.sunbirded.org/api/framework/v1/category/read/topic?framework=<FRAMEWORK_ID>	
	3. create terms using category ID and framework ID. using following End point https://dev.sunbirded.org/api/framework/v1/term/update?framework=<FRAMEWORK_ID&category=<CATEGORY_ID>
	
### **API Endpoints used in create/upload flow**
        1. https://dock.sunbirded.org/api/channel/v1/read/<CHANNEL_ID>
        2. https://dev.sunbirded.org/api/framework/v1/category/read/topic?framework=<FRAMEWORK_ID>	
        3. https://dev.sunbirded.org/api/framework/v1/term/update?framework=<FRAMEWORK_ID&category=<CATEGORY_ID>
        
### **CURL**
`curl --location --request POST 'https://dev.sunbirded.org/api/framework/v1/term/create?framework=<FRAMEWORK_ID>&category=<CATEGORY_ID>' \
--header 'Content-Type: application/json' \
--header '<TOKEN> \
--header 'X-Channel-Id: <CHANNEL_ID>' \
--data '{
	"request": {
		"term": {
            "code": "Other Curriculum Subjects",
            "name": "Other Curriculum Subjects",
            "label":"Other Curriculum Subjects",
            "description": "Other Curriculum Subjects",
            "children": [
              {
                "code": "ST_Physical Education",
                "name": "Physical Education",
                "label": "Physical Education",
                "description": "Physical Education"
              },
              {
                "code": "ST_Economics",
                "name": "Economics",
                "label": "Economics",
                "description": "Economics"
              },
              {
                "code": "ST_Mathematics",
                "name": "Mathematics",
                "label": "Mathematics",
                "description": "Mathematics"
              },
              {
                "code": "ST_Accountancy",
                "name": "Accountancy",
                "label": "Accountancy",
                "description": "Accountancy"
              },
              {
                "code": "ST_Business Studies",
                "name": "Business Studies",
                "label": "Business Studies",
                "description": "Business Studies"
              },
              {
                "code": "ST_Psychology",
                "name": "Psychology",
                "label": "Psychology",
                "description": "Psychology"
              },
              {
                "code": "ST_Education",
                "name": "Education",
                "label": "Education",
                "description": "Education"
              }
            ]
          }
	}
}
'`
### Error Response 
`{
    "id": "api.term.create",
    "ver": "1.0",
    "ts": "2024-01-19T03:50:10.528Z",
    "params": {
        "resmsgid": "d7f35600-b67d-11ee-b7cc-a7f3aa45b5e1",
        "msgid": null,
        "status": "failed",
        "err": "ERR_SERVER_ERROR",
        "errmsg": "Internal Server Error"
    },
    "responseCode": "SERVER_ERROR",
    "result": {}
}`
### Additional Information
Reference document for Framework creation : https://project-sunbird.atlassian.net/l/cp/DQ8cAQVX
Tested on Google Chrome: Version 119.0.6045.199 (Official Build) (arm64)

can someone help on this.
cc: @kiranharidas187

**Comment by pallakartheekreddy** [comment]
@ramkumar-pacewisdom In the above request, you are trying to create child terms as part of parent teams creation, which is wrong. Please create child teams independently using the same API and add them as children as shown below request.

**Step 1**:  Create child terms first

Create child teams using the below Request
```
URL: https://dev.sunbirded.org/api/framework/v1/term/create?framework=<FRAMEWORK_ID>&category=<CATEGORY_ID>
Request:
{
  "request": {
    "term": {
      "code": "ST_Physical Education",
      "name": "Physical Education",
      "label": "Physical Education",
      "description": "Physical Education"
    }
  }
}
```

**Step 2**: Create parent term and add childerns to it.
```
URL: https://dev.sunbirded.org/api/framework/v1/term/create?framework=<FRAMEWORK_ID>&category=<CATEGORY_ID>
Request:
{
    "request": {
        "term": {
            "code": "Other_Curriculum_Subjects",
            "name": "Other Curriculum Subjects",
            "label": "Other Curriculum Subjects",
            "description": "Other Curriculum Subjects",
            "children": [
                {
                    "identifier": "< FRAMEWORK_ID>_< CATEGORY_ID>_st_physical-education"
                },
                {
                    "identifier": "< FRAMEWORK_ID>_< CATEGORY_ID>_st_economics"
                },
                {
                    "identifier": "< FRAMEWORK_ID>_< CATEGORY_ID>_st_mathematics"
                }
            ]
        }
    }
}
```

**Reply by ramkumar-ns-dev** [reply]
@pallakartheekreddy I've followed the steps you've mentioned above, but it's throws following error
`{
    "id": "api.term.create",
    "ver": "1.0",
    "ts": "2024-01-23T07:36:35.358Z",
    "params": {
        "resmsgid": "22cae3e0-b9c2-11ee-aa5c-83967cdd479c",
        "msgid": null,
        "status": "failed",
        "err": "ERR_TERM_CODE_REQUIRED",
        "errmsg": "Unique code is required for Term"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {}
}`

**I added code property, Then it's returning same 500 internal server error**

`{
    "id": "api.term.create",
    "ver": "1.0",
    "ts": "2024-01-23T07:37:45.341Z",
    "params": {
        "resmsgid": "4c8172d0-b9c2-11ee-aa5c-83967cdd479c",
        "msgid": null,
        "status": "failed",
        "err": "ERR_SERVER_ERROR",
        "errmsg": "Internal Server Error"
    },
    "responseCode": "SERVER_ERROR",
    "result": {}
}`

**Reply by pallakartheekreddy** [reply]
@ramkumar-pacewisdom Can you share the request?

**Reply by ramkumar-ns-dev** [reply]
@pallakartheekreddy payload for creating terms

1. `curl --location --request POST 'https://dev.sunbirded.org/api/framework/v1/term/create?framework=TPD&category=topic' \
--header 'Content-Type: application/json' \
--header 'Authorization: <token>' \
--header 'X-Channel-Id: 01379157488516300825' \
--data '{
    "request": {
        "term": {
            "code": "ICT",
            "label": "ICT",
            "name": "ICT",
            "description": "ICT"
        }
    }
}'`

**response for above request is Status 200**

2. payload for creation children under terms
`curl --location --request POST 'https://dev.sunbirded.org/api/framework/v1/term/create?framework=TPD&category=topic' \
--header 'Content-Type: application/json' \
--header 'Authorization: <TOKEN>' \
--header 'X-Channel-Id: 01379157488516300825' \
--data '{
    "request": {
        "term": {
            "code": "ICT",
            "label": "ICT",
            "name": "ICT",
            "description": "ICT",
            "children": [
              {
                "identifier": "tpd_topic_ict"
              }
            ]
        }
    }
}'`

**response for above request**

`{
    "id": "api.term.create",
    "ver": "1.0",
    "ts": "2024-01-23T08:59:37.117Z",
    "params": {
        "resmsgid": "bc26fcd0-b9cd-11ee-aa5c-83967cdd479c",
        "msgid": null,
        "status": "failed",
        "err": "ERR_TERM_CODE_REQUIRED",
        "errmsg": "Unique code is required for Term"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {}
}`

3. I sent request with Code property.
`curl --location --request POST 'https://dev.sunbirded.org/api/framework/v1/term/create?framework=TPD&category=topic' \
--header 'Content-Type: application/json' \
--header 'Authorization: <TOKEN>' \
--header 'X-Channel-Id: 01379157488516300825' \
--data '{
    "request": {
        "term": {
            "code": "ICT",
            "label": "ICT",
            "name": "ICT",
            "description": "ICT",
            "children": [
              {
                "identifier": "tpd_topic_ict",
                "code": "ICT"
              }
            ]
        }
    }
}'`

**response i got 500 error**

`{
    "id": "api.term.create",
    "ver": "1.0",
    "ts": "2024-01-23T09:02:11.553Z",
    "params": {
        "resmsgid": "18340d10-b9ce-11ee-aa5c-83967cdd479c",
        "msgid": null,
        "status": "failed",
        "err": "ERR_SERVER_ERROR",
        "errmsg": "Internal Server Error"
    },
    "responseCode": "SERVER_ERROR",
    "result": {}
}`

**Reply by pallakartheekreddy** [reply]
@ramkumar-pacewisdom did you see the error message properly? You are using the same `code`(ICT) in both requests, which causing the issue and error message clearly saying `Unique code is required for Term`. Please consider using a different code(Unique/different code for each term), the issue will be resolved.

**Reply by amitpriyadarshi** [reply]
@ramkumar-pacewisdom in the below term create request:
`{ "request": { "term": { "code": "ICT", "label": "ICT", "name": "ICT", "description": "ICT", "children": [ { "identifier": "tpd_topic_ict" } ] } } }`
You are trying to create term with code: `ICT` which is already existing as `tpd_topic_ict`. Which is wrong.
Also, you should not pass `tpd_topic_ict` as children while creating `tpd_topic_ict`. 

The second request is not correct. Code should not be passed within children. Because in children we should pass only identifier.
cc: @pallakartheekreddy

**Reply by ramkumar-ns-dev** [reply]
@amitpriyadarshi @pallakartheekreddy  got it, Really thanks for the response. I'm trying to call this API for multiple time but not working in mine, I will do the changes and update here once done.

**Reply by ramkumar-ns-dev** [reply]
@amitpriyadarshi @pallakartheekreddy Now i'm able to add children property to a term by passing another terms as a children and it's working. thank you.
